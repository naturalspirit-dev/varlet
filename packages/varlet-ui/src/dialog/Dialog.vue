<template>
	<var-popup
		class="var-dialog__popup-radius"
		v-model:show="popupShow"
		:overlay="overlay"
		:overlay-class="overlayClass"
		:lock-scroll="lockScroll"
		:close-on-click-overlay="popupCloseOnClickOverlay"
		:teleport="teleport"
		@open="$props.onOpen"
		@close="$props.onClose"
		@opened="$props.onOpened"
		@closed="$props.onClosed"
		@click-overlay="handleClickOverlay"
	>
		<div class="var--box var-dialog" v-bind="$attrs">
			<div class="var-dialog__title">
				<slot name="title">{{ title }}</slot>
			</div>
			<div
				class="var-dialog__message"
				:style="{
					'text-align': messageAlign,
				}"
			>
				<slot>
					{{ message }}
				</slot>
			</div>
			<div class="var-dialog__actions">
				<var-button
					class="var-dialog__button"
					:color="cancelButtonColor"
					:background="cancelButtonBackground"
					v-if="cancelButton"
					@click="cancel"
				>
					{{ cancelButtonText }}
				</var-button>
				<var-button
					class="var-dialog__button"
					:color="confirmButtonColor"
					:background="confirmButtonBackground"
					v-if="confirmButton"
					@click="confirm"
				>
					{{ confirmButtonText }}
				</var-button>
			</div>
		</div>
	</var-popup>
</template>

<script lang="ts">
import Popup from '../popup'
import Button from '../button'
import { props } from './props'
import { defineComponent, ref, Ref, watch } from 'vue'

export default defineComponent({
	name: 'VarDialog',
	components: {
		[Popup.name]: Popup,
		[Button.name]: Button,
	},
	inheritAttrs: false,
	props,
	setup(props) {
		// 需要透传到popup组件里并需要特殊处理的参数
		const popupShow: Ref<boolean> = ref(false)
		const popupCloseOnClickOverlay: Ref<boolean> = ref(false)

		watch(
			() => props.show,
			(newValue) => {
				popupShow.value = newValue
			},
			{ immediate: true }
		)
		watch(
			() => props.closeOnClickOverlay,
			(newValue) => {
				if (props.onBeforeClose) {
					// 异步关闭情况下 禁止点击弹出层关闭
					popupCloseOnClickOverlay.value = false
					return
				}
				popupCloseOnClickOverlay.value = newValue
			},
			{ immediate: true }
		)

		// 异步关闭回调
		const done = () => props['onUpdate:show']?.(false)

		const handleClickOverlay = () => {
			props.onClickOverlay?.()
			if (props.closeOnClickOverlay === false) {
				return
			}
			if (props.onBeforeClose) {
				props.onBeforeClose(done)
				return
			}
			props['onUpdate:show']?.(false)
		}

		const confirm = () => {
			props.onConfirm?.()
			if (props.beforeClose) {
				props.onBeforeClose(done)
				return
			}
			props['onUpdate:show']?.(false)
		}

		const cancel = () => {
			props.onCancel?.()
			if (props.onBeforeClose) {
				props.onBeforeClose(done)
				return
			}
			props['onUpdate:show']?.(false)
		}

		return {
			popupShow,
			popupCloseOnClickOverlay,
			handleClickOverlay,
			confirm,
			cancel,
		}
	},
})
</script>

<style lang="less">
@import './dialog';
</style>
