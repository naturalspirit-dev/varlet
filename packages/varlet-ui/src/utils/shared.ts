export interface CacheInstance<T> {
	cache: T[]

	has(key: T): boolean

	add(key: T): void

	remove(key: T): void
}

export const isString = (val: unknown): val is string => typeof val === 'string'

export const isNumber = (val: unknown): val is number => typeof val === 'number'

export const isBaseObject = (val: unknown): val is Record<string, any> =>
  Object.prototype.toString.call(val) === '[object Object]'

export const isArray = (val: unknown): val is Array<any> => Array.isArray(val)

export const isURL = (val: string) => /^(http)|(\.*\/)/.test(val)

export const isEmpty = (val: unknown) =>
  val === undefined || val === null || val === '' || (Array.isArray(val) && !val.length)

export const removeItem = (arr: Array<unknown>, item: unknown) => {
  if (arr.length) {
    const index: number = arr.indexOf(item)
    if (index > -1) {
      return arr.splice(index, 1)
    }
  }
}

export const throttle = (method: any, mustRunDelay = 200): (() => void) => {
  let timer: number
  let start: number
  return function loop(this: unknown, ...args) {
    const now = Date.now()
    if (!start) {
      start = now
    }
    if (timer) {
      window.clearTimeout(timer)
    }
    if (now - start >= mustRunDelay) {
      method.apply(this, args)
      start = now
    } else {
      timer = window.setTimeout(() => {
        loop.apply(this, args)
      }, 50)
    }
  }
}

export const debounce = (method: any, delay = 200) => {
  let timer: number
  return function (this: unknown, ...args: any[]) {
    timer && window.clearTimeout(timer)
    timer = window.setTimeout(() => {
      method.apply(this, args)
    }, delay)
  }
}

export const checkIntersectionObserverAPI = (): boolean => 'IntersectionObserver' in window

export const createInViewportObserver = <T>(handler: (el: T) => void): IntersectionObserver => {
  return new IntersectionObserver((intersectionObserverEntries: IntersectionObserverEntry[]) => {
    intersectionObserverEntries.forEach((intersectionObserverEntry: IntersectionObserverEntry) => {
      if (intersectionObserverEntry.intersectionRatio > 0) {
        handler((intersectionObserverEntry.target as unknown) as T)
      }
    })
  })
}

export const createCache = <T>(max: number): CacheInstance<T> => {
  const cache: T[] = []

  return {
    cache,
    has(key: T) {
      return cache.includes(key)
    },
    add(key: T) {
      if (this.has(key)) {
        return
      }

      this.cache.length === max && cache.shift()
      this.cache.push(key)
    },
    remove(key: T) {
      this.has(key) && removeItem(this.cache, key)
    },
  }
}
