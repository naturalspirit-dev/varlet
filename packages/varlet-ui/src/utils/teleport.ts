import { onActivated, onDeactivated, ref, Ref } from 'vue'

export function useTeleport() {
  const disabled: Ref<boolean> = ref(false)
  onActivated(() => {
    disabled.value = false
  })
  onDeactivated(() => {
    disabled.value = true
  })

  return {
    disabled
  }
}
