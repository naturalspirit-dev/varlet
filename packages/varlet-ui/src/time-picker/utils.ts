import { hours24, hoursAmpm } from './props'
import type { AmPm } from './props'
import { toNumber } from '../utils/shared'

type DisableProps = {
  time: number
  format: string
  ampm: AmPm | undefined
  hour: string
  minute: number
  max: string | undefined
  min: string | undefined
  disableHour: Array<string>
}

export const notConvert = (format: string, ampm: AmPm | undefined): boolean => format === '24hr' || ampm === 'am'

export const convertHour = (format: string, ampm: AmPm | undefined, hour: string) => {
  const index = hoursAmpm.findIndex((hourAmpm) => toNumber(hourAmpm) === toNumber(hour))
  const getHour = notConvert(format, ampm) ? hour : hours24[index]

  return {
    hourStr: getHour,
    hourNum: toNumber(getHour),
  }
}

export const getNumberTime = (time: string) => {
  const [hour, minute, second] = time.split(':')

  return {
    hour: toNumber(hour),
    minute: toNumber(minute),
    second: toNumber(second),
  }
}

export const getIsDisableMinute = (values: Omit<DisableProps, 'minute'>): boolean => {
  const { time, format, ampm, hour, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    return maxHour === hourNum && time > maxMinute
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    return minHour === hourNum && time < minMinute
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    return (minHour === hourNum && time < minMinute) || (maxHour === hourNum && time > maxMinute)
  }
  return false
}

export const getIsDisableSecond = (values: DisableProps): boolean => {
  const { time, format, ampm, hour, minute, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)

    return (maxHour === hourNum && maxMinute < minute) || (maxMinute === minute && time > maxSecond)
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    return (minHour === hourNum && minMinute > minute) || (minMinute === minute && time > minSecond)
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    return (
      (maxHour === hourNum && maxMinute < minute) ||
      (minHour === hourNum && minMinute > minute) ||
      (maxHour === hourNum && maxMinute === minute && time > maxSecond) ||
      (minHour === hourNum && minMinute === minute && time < minSecond)
    )
  }
  return false
}
