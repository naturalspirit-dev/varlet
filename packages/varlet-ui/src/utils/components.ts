import {
	createApp,
	Component,
	h,
	getCurrentInstance,
	inject,
	onUnmounted,
	computed,
	ComputedRef,
	provide,
	reactive,
	VNode,
	isVNode,
	ComponentInternalInstance,
	onMounted,
	onBeforeUnmount,
	nextTick,
} from 'vue'
import { removeItem } from './shared'

export interface MountInstance {
	instance: any
	unmount: () => void
}

export interface ChildrenCounter {
	collect(instance: ComponentInternalInstance): void
	clear(instance: ComponentInternalInstance): void
	instances: ComponentInternalInstance[]
}

export interface BaseParentProvider<C> {
	collect(childProvider: C): void
	clear(childProvider: C): void
}

export function pickProps(props: any, propsKey: string): any
export function pickProps(props: any, propsKey: string[]): any
export function pickProps(props: any, propsKey: any): any {
	return Array.isArray(propsKey)
		? propsKey.reduce((pickedProps: any, key) => {
				pickedProps[key] = props[key]
				return pickedProps
		  }, {})
		: props[propsKey]
}

export function mount(component: Component): MountInstance {
	const app = createApp(component)
	const host = document.createElement('div')
	document.body.appendChild(host)

	return {
		instance: app.mount(host),
		unmount() {
			app.unmount(host)
			document.body.removeChild(host)
		},
	}
}

export function mountInstance(
	component: any,
	props: Record<string, any> = {},
	eventListener: Record<string, any> = {}
): {
	unmountInstance: () => void
} {
	const Host = {
		setup() {
			return () =>
				h(component, {
					...props,
					...eventListener,
				})
		},
	}

	const { unmount } = mount(Host)
	return { unmountInstance: unmount }
}

export function flatVNodes(subTree: any) {
	const vNodes: VNode[] = []

	const flat = (subTree: any) => {
		if (subTree?.component) {
			flat(subTree?.component.subTree)
			return
		}

		if (Array.isArray(subTree?.children)) {
			subTree.children.forEach((child: any) => {
				if (isVNode(child)) {
					vNodes.push(child)

					flat(child)
				}
			})
		}
	}

	flat(subTree)

	return vNodes
}

export function useAtChildrenCounter(key: symbol) {
	const instances: ComponentInternalInstance[] = reactive([])
	const parentInstance: ComponentInternalInstance = getCurrentInstance() as ComponentInternalInstance

	const sortInstances = () => {
		const vNodes: any[] = flatVNodes(parentInstance.subTree)

		instances.sort((a, b) => {
			return vNodes.indexOf(a.vnode) - vNodes.indexOf(b.vnode)
		})
	}

	const collect = (instance: ComponentInternalInstance) => {
		instances.push(instance)
		sortInstances()
	}

	const clear = (instance: ComponentInternalInstance) => {
		removeItem(instances, instance)
	}

	provide<ChildrenCounter>(key, {
		collect,
		clear,
		instances,
	})

	const length: ComputedRef<number> = computed(() => instances.length)

	return {
		length,
	}
}

export function useAtParentIndex(key: symbol) {
	const childrenCounter: ChildrenCounter | undefined = inject<ChildrenCounter>(key)
	if (!childrenCounter) {
		throw new Error('Children counter provider not found')
	}

	const { collect, clear, instances } = childrenCounter

	const instance: ComponentInternalInstance = getCurrentInstance() as ComponentInternalInstance

	nextTick().then(() => collect(instance))
	onUnmounted(() => clear(instance))

	const index = computed(() => instances.indexOf(instance))

	return {
		index,
	}
}

export function useChildren<P, C>(key: symbol) {
	const childProviders: C[] = []

	const collect = (childProvider: C) => {
		childProviders.push(childProvider)
	}

	const clear = (childProvider: C) => {
		removeItem(childProviders, childProvider)
	}

	const bindChildren = (parentProvider: P) => {
		provide<P & BaseParentProvider<C>>(key, {
			collect,
			clear,
			...parentProvider,
		})
	}

	return {
		childProviders,
		bindChildren,
	}
}

export function useParent<P, C>(key: symbol) {
	const { collect, clear, ...parentProvider } = inject<P & BaseParentProvider<C>>(key) as P & BaseParentProvider<C>
	if (!collect || !clear) {
		throw new Error('Parent not found')
	}

	const bindParent = (childProvider: C) => {
		onMounted(() => collect(childProvider))
		onBeforeUnmount(() => clear(childProvider))
	}

	return {
		parentProvider,
		bindParent,
	}
}
