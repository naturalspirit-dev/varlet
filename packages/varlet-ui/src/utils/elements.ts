import { isNumber, isString, toNumber } from './shared'

export function getTop(element: HTMLElement): number {
  const { top } = element.getBoundingClientRect()

  return top + (document.body.scrollTop || document.documentElement.scrollTop)
}

export function getScrollTop(element: Element | Window): number {
  const top = 'scrollTop' in element ? element.scrollTop : element.pageYOffset

  // iOS scroll bounce cause minus scrollTop
  return Math.max(top, 0)
}

export function getScrollLeft(element: Element | Window): number {
  const left = 'scrollLeft' in element ? element.scrollLeft : element.pageXOffset

  return Math.max(left, 0)
}

export function getLeft(element: HTMLElement): number {
  const { left } = element.getBoundingClientRect()
  return left + (document.body.scrollLeft || document.documentElement.scrollLeft)
}

export function inViewport(element: HTMLElement): boolean {
  const { top, bottom, left, right } = element.getBoundingClientRect()

  const xInViewport = left < window.innerWidth && right > 0
  const yInViewport = top < window.innerHeight && bottom > 0

  return xInViewport && yInViewport
}

export function getTranslate(el: HTMLElement) {
  const { transform } = window.getComputedStyle(el)
  return +transform.slice(transform.lastIndexOf(',') + 2, transform.length - 1)
}

export function isHidden(el: HTMLElement) {
  const { width, height } = el.getBoundingClientRect()

  return width === 0 && height === 0
}

export function getParentScroller(el: HTMLElement): HTMLElement | Window {
  let element = el

  while (element) {
    if (!element.parentNode) {
      break
    }

    element = element.parentNode as HTMLElement

    if (element === document.body || element === document.documentElement) {
      break
    }

    const scrollRE = /(scroll|auto)/
    const { overflowY } = window.getComputedStyle(element)

    if (scrollRE.test(overflowY)) {
      return element
    }
  }

  return window
}

// example 1rem
export const isRem = (value: unknown) => isString(value) && value.endsWith('rem')

// example 1 || 1px
export const isPx = (value: unknown) => (isString(value) && value.endsWith('px')) || isNumber(value)

// example 1%
export const isPercent = (value: unknown) => isString(value) && value.endsWith('%')

// example 1vw
export const isVw = (value: unknown) => isString(value) && value.endsWith('vw')

// example 1vh
export const isVh = (value: unknown) => isString(value) && value.endsWith('vh')

// example return 1
export const toPxNum = (value: unknown): number => {
  if (isNumber(value)) {
    return value
  }

  if (isPx(value)) {
    return +(value as string).replace('px', '')
  }

  if (isVw(value)) {
    return (+(value as string).replace('vw', '') * window.innerWidth) / 100
  }

  if (isVh(value)) {
    return (+(value as string).replace('vh', '') * window.innerHeight) / 100
  }

  if (isRem(value)) {
    const num = +(value as string).replace('rem', '')
    const rootFontSize = window.getComputedStyle(document.documentElement).fontSize

    return num * parseFloat(rootFontSize)
  }

  if (isString(value)) {
    return toNumber(value)
  }

  // % and other
  return 0
}

// example return 1px 1% 1vw 1vh 1rem null
export const toSizeUnit = (value: unknown) => {
  if (value == null) {
    return null
  }

  if (isPercent(value) || isVw(value) || isVh(value) || isRem(value)) {
    return value
  }

  return `${toPxNum(value)}px`
}

export function requestAnimationFrame(fn: FrameRequestCallback): number {
  return window.requestAnimationFrame ? window.requestAnimationFrame(fn) : window.setTimeout(fn, 16)
}

export function cancelAnimationFrame(handle: number): void {
  window.cancelAnimationFrame ? window.cancelAnimationFrame(handle) : window.clearTimeout(handle)
}

export function nextTickFrame(fn: FrameRequestCallback) {
  requestAnimationFrame(() => {
    requestAnimationFrame(fn)
  })
}

interface ScrollToOptions {
  top?: number
  left?: number
  duration?: number
  animation: (progress: number) => number
}

export function scrollTo(
  element: HTMLElement | Window,
  { top = 0, left = 0, duration = 300, animation }: ScrollToOptions
) {
  const startTime = Date.now()

  const scrollTop = getScrollTop(element)
  const scrollLeft = getScrollLeft(element)

  const frame = () => {
    const progress = (Date.now() - startTime) / duration

    if (progress < 1) {
      const nextTop = scrollTop + (top - scrollTop) * animation(progress)
      const nextLeft = scrollLeft + (left - scrollLeft) * animation(progress)

      element.scrollTo(nextLeft, nextTop)
      requestAnimationFrame(frame)
    } else {
      element.scrollTo(left, top)
    }
  }

  requestAnimationFrame(frame)
}
