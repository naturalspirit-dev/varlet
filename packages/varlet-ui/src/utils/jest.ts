import { nextTick } from 'vue'
import type { ComponentPublicInstance } from 'vue'
import { config } from '@vue/test-utils'
import type { VueWrapper, DOMWrapper } from '@vue/test-utils'
import { isPlainObject } from './shared'

export const delay = (time: number) => new Promise((resolve) => setTimeout(resolve, time))

export function getTouch(el: Element | Window, x: number, y: number) {
  return {
    identifier: Date.now(),
    target: el,
    pageX: x,
    pageY: y,
    clientX: x,
    clientY: y,
    radiusX: 2.5,
    radiusY: 2.5,
    rotationAngle: 10,
    force: 0.5,
  }
}

export function trigger(
  wrapper: VueWrapper<ComponentPublicInstance<any, any, any>> | DOMWrapper<Element> | Element | Window,
  eventName: string,
  x = 0,
  y = 0,
  offsetX = 0,
  offsetY = 0
) {
  const el = 'element' in wrapper ? wrapper.element : wrapper
  const touchList = [getTouch(el, x, y)]

  const event = document.createEvent('CustomEvent')
  event.initCustomEvent(eventName, true, true, {})

  Object.assign(event, {
    clientX: x,
    clientY: y,
    offsetX,
    offsetY,
    touches: touchList,
    targetTouches: touchList,
    changedTouches: touchList,
  })

  el.dispatchEvent(event)

  return nextTick()
}

export function mockOffset() {
  Object.defineProperties(HTMLElement.prototype, {
    offsetWidth: {
      get() {
        return parseFloat(window.getComputedStyle(this).width) || 0
      },
    },
    offsetHeight: {
      get() {
        return parseFloat(window.getComputedStyle(this).height) || 0
      },
    },
    clientWidth: {
      get() {
        return parseFloat(window.getComputedStyle(this).width) || 0
      },
    },
    clientHeight: {
      get() {
        return parseFloat(window.getComputedStyle(this).height) || 0
      },
    },
    offsetLeft: {
      get() {
        return parseFloat(window.getComputedStyle(this).marginLeft) || 0
      },
    },
    offsetTop: {
      get() {
        return parseFloat(window.getComputedStyle(this).marginTop) || 0
      },
    },
    offsetParent: {
      get() {
        return this.parentNode ?? {}
      },
    },
  })
}

interface Entry {
  intersectionRatio: number
  target: null | HTMLElement
}

const entry: Entry = {
  intersectionRatio: 0,
  target: null,
}

let callback: any

export function mockIntersectionObserver() {
  const originMethod = window.IntersectionObserver

  const IntersectionObserver = function (cb: any) {
    callback = cb
  }
  IntersectionObserver.prototype.observe = (el: HTMLElement) => {
    entry.target = el
  }
  IntersectionObserver.prototype.unobserve = (el: HTMLElement) => {
    entry.target = null
  }

  Object.assign(window, {
    IntersectionObserver,
  })

  return {
    mockRestore() {
      entry.intersectionRatio = 0
      entry.target = null
      window.IntersectionObserver = originMethod
    },
    trigger(intersectionRatio: number) {
      entry.intersectionRatio = intersectionRatio
      callback?.([entry])
      return entry.target
    },
  }
}

export async function triggerDrag(
  el: VueWrapper<ComponentPublicInstance<any, any, any>> | DOMWrapper<Element> | HTMLElement,
  x = 0,
  y = 0
) {
  await trigger(el, 'touchstart', 0, 0)
  await trigger(el, 'touchmove', x / 4, y / 4)
  await trigger(el, 'touchmove', x / 3, y / 3)
  await trigger(el, 'touchmove', x / 2, y / 2)
  await trigger(el, 'touchmove', x, y)
  await trigger(el, 'touchend', x, y)
}

export function mockTranslate() {
  const originMethod = window.getComputedStyle

  const XRE = /translateX\((\d+)px\)/
  const YRE = /translateY\((\d+)px\)/

  Object.assign(window, {
    getComputedStyle: (el: HTMLElement) => {
      const styles = originMethod.call(window, el)
      const x = styles.transform.match(XRE)?.[1] ?? 0
      const y = styles.transform.match(YRE)?.[1] ?? 0
      styles.transform = `matrix(1, 0, 0, 1, ${x}, ${y})`
      return styles
    },
  })

  return {
    mockRestore() {
      window.getComputedStyle = originMethod
    },
  }
}

interface MockFileReaderCtx {
  result: string
  onload: () => void
  readAsDataURL: () => void
}

export function mockFileReader(url: string) {
  const originMethod = window.FileReader

  Object.assign(window, {
    FileReader(this: MockFileReaderCtx) {
      this.onload = function () {}
      this.result = ''

      this.readAsDataURL = function () {
        this.result = url
        this.onload()
      }
    },
  })

  return {
    mockRestore() {
      window.FileReader = originMethod
    },
  }
}

export function mockStubs() {
  const originStubs = config.global.stubs

  config.global.stubs = {}

  return {
    mockRestore() {
      config.global.stubs = originStubs
    },
  }
}

export function mockConsole(method: keyof Console) {
  const originMethod = console[method]

  console[method] = () => {}

  return {
    mockRestore() {
      console[method] = originMethod
    },
  }
}

export function mockScrollTo(Element: any) {
  Element.prototype.scrollTo = function (x: ScrollToOptions | number, y?: number) {
    if (isPlainObject(x)) {
      this.scrollLeft = x.left
      this.scrollTop = x.top
    } else {
      this.scrollLeft = x
      this.scrollTop = y
    }
  }
}
