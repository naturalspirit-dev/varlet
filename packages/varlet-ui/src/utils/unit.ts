import { isNumber, isString } from './shared'


