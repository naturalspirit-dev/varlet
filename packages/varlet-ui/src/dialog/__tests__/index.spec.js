import example from '../example'
import Dialog from '..'
import VarDialog from '../Dialog'
import { mount } from '@vue/test-utils'
import { createApp } from 'vue'
import { delay, trigger } from '../../utils/jest'

test('test dialog example', () => {
  const wrapper = mount(example)
  expect(wrapper.html()).toMatchSnapshot()
  wrapper.unmount()
})

test('test dialog plugin', () => {
  const app = createApp({}).use(Dialog)
  expect(app.component(VarDialog.name)).toBeTruthy()
})

test('test dialog functional show & close', async () => {
  const onOpen = jest.fn()
  const onClose = jest.fn()
  Dialog({
    message: 'test message',
    onOpen,
    onClose,
  })

  await delay(16)
  expect(onOpen).toHaveBeenCalledTimes(1)
  expect(document.querySelector('.var-popup').style.display).toBe('')

  Dialog.close()

  await delay(16)
  expect(onClose).toHaveBeenCalledTimes(1)
  expect(document.querySelector('.var-popup').style.display).toBe('none')
})

test('test dialog functional confirm & cancel', async () => {
  const onConfirm = jest.fn()
  const onCancel = jest.fn()

  Dialog({
    message: 'test confirm',
    onConfirm,
  })
  await delay(16)
  await trigger(document.querySelectorAll('.var-dialog__confirm-button')[1], 'click')
  expect(onConfirm).toHaveBeenCalledTimes(1)

  Dialog({
    message: 'test cancel',
    onCancel,
  })
  await delay(16)
  await trigger(document.querySelectorAll('.var-dialog__cancel-button')[2], 'click')
  expect(onCancel).toHaveBeenCalledTimes(1)
})
