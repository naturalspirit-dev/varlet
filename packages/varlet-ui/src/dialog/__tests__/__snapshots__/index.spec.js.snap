// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test dialog example 1`] = `
"<div class=\\"app-type\\">函数调用</div><button class=\\"var-button var--box var-button--normal var--flex var-button--primary var-elevation--1\\" style=\\"width: 100%;\\">
  <!--v-if-->
  <div class=\\"var-button__content\\">基本使用</div>
</button><button class=\\"var-button var--box var-button--normal var--flex var-button--primary var-elevation--1\\" style=\\"width: 100%;\\">
  <!--v-if-->
  <div class=\\"var-button__content\\">修改标题</div>
</button><button class=\\"var-button var--box var-button--normal var--flex var-button--primary var-elevation--1\\" style=\\"width: 100%;\\">
  <!--v-if-->
  <div class=\\"var-button__content\\">隐藏按钮</div>
</button><button class=\\"var-button var--box var-button--normal var--flex var-button--primary var-elevation--1\\" style=\\"width: 100%;\\">
  <!--v-if-->
  <div class=\\"var-button__content\\">处理用户行为</div>
</button><button class=\\"var-button var--box var-button--normal var--flex var-button--primary var-elevation--1\\" style=\\"width: 100%;\\">
  <!--v-if-->
  <div class=\\"var-button__content\\">异步关闭</div>
</button>
<div class=\\"app-type\\">组件调用</div><button class=\\"var-button var--box var-button--normal var--flex var-button--primary var-elevation--1\\" style=\\"width: 100%;\\">
  <!--v-if-->
  <div class=\\"var-button__content\\">基本使用</div>
</button>
<!--teleport start-->
<!--teleport end--><button class=\\"var-button var--box var-button--normal var--flex var-button--primary var-elevation--1\\" style=\\"width: 100%;\\">
  <!--v-if-->
  <div class=\\"var-button__content\\">异步关闭</div>
</button>
<!--teleport start-->
<!--teleport end--><button class=\\"var-button var--box var-button--normal var--flex var-button--primary var-elevation--1\\" style=\\"width: 100%;\\">
  <!--v-if-->
  <div class=\\"var-button__content\\">自定义插槽</div>
</button>
<!--teleport start-->
<!--teleport end-->"
`;
